### PR TITLE
[WIP] do not omit image references that are not filtered on

### LINKED ISSUE #29203: Docker Repository API to get images with filter returns "RepoDigests" as null
Docker Repository API to get images with filter returns "RepoDigests" as null.

**Steps to reproduce the issue:**
1. Call Docker API on a repository without a filter (i.e. `.../images/json`)
2. Call Docker API on a repository with a filter (e.g. `.../images/json?filter=<anImageName>`)
3. Compare the returned "RepoDigests" fields of the first response with that of the second response

**Describe the results you received:**
The API responds with a null "RepoDigests" when a filter is used

**Describe the results you expected:**
I expect the API to return the same response whether I use the filter or not

**Additional information you deem important (e.g. issue happens only occasionally):**
N/A

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Thu Oct 27 00:09:21 2016
 OS/Arch:      windows/amd64
 Experimental: true

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Thu Oct 27 00:09:21 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 26
 Running: 25
 Paused: 0
 Stopped: 1
Images: 26
Server Version: 1.12.3
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay null host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.27-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.775 GiB
Name: moby
ID: Z26R:GZO7:W4TE:77XS:6XQT:XS7R:ABTC:DFAA:LJLL:AR7W:QIPL:HBK6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
Experimental: true
Insecure Registries:
 md1drepo01.lnx.ix.com:5000
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
None

### LINKED ISSUE #29901: Digest not shown when image tag specified
Digests are shown for pushed images with `docker images --digests ...`

```
$ docker push example.com/foo/bar:master
The push refers to a repository [example.com/foo/bar]
9ce98538809e: Layer already exists 
91a8f3990fa4: Layer already exists 
837094365fb4: Layer already exists 
9708155555f8: Layer already exists 
b91611bea4a8: Layer already exists 
148b5a1af9ed: Layer already exists 
011b303988d2: Layer already exists 
master: digest: sha256:164b5fc83e34c2c1d59a430c7b64ecd010902f13c62ba9dcc50efb771c4060b2 size: 12564
                                                                                                                                                              
$ docker images --digests example.com/foo/bar
REPOSITORY                         TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
example.com/foo/bar   master              sha256:164b5fc83e34c2c1d59a430c7b64ecd010902f13c62ba9dcc50efb771c4060b2   d594640ae3af        2 hours ago         63.02 MB
example.com/foo/bar   <none>              sha256:6fd65f7141dc61adeb9f262784b1aca9848866d2d4b68c3a10ecf0aa2d1edef6   1c270a40466d        3 days ago          63.02 MB
example.com/foo/bar   <none>              sha256:0b9e94db5aaedb6f6407778d803e99b0e7dd2e710842611873411717dd51e7b8   c42d40e9c60b        3 days ago          63.02 MB
example.com/foo/bar   <none>              sha256:a5aee68356dba08f0576396d8cbd23cf48630391e00d3197c2b237cbedc8a7a2   998b383b2bcb        2 weeks ago         63.02 MB
example.com/foo/bar   <none>              sha256:076ff9b7dc7cadf2a74d7bf5b7226c0b69ef18e9d29834c2c2e0f3efca019aba   a6df32f2ad19        2 weeks ago         62.08 MB
example.com/foo/bar   <none>              sha256:a8a4411a3f6f89c9215a24adfa5ae3c62d04052c53039fd594002a78118ecf1d   7bc456afeef5        2 weeks ago         62.08 MB
example.com/foo/bar   <none>              sha256:eb448ad5b5d1a6b8d425e90c37056c82c90809ffbf3e6486f601351893fe2785   973e019e1f45        4 weeks ago         429.8 MB
```
But not when you specify an image tag (a bug).
```                                                                                                                                                              
$ docker images --digests example.com/foo/bar:master
REPOSITORY                         TAG                 DIGEST              IMAGE ID            CREATED             SIZE
example.com/foo/bar   master              <none>              d594640ae3af        2 hours ago         63.02 MB
```

```
$ docker version
Client:
 Version:      1.12.2-rc1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   45bed2c
 Built:        Tue Sep 27 23:38:15 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 22:01:48 2016
 OS/Arch:      linux/amd64

$ docker info
Containers: 28
 Running: 9
 Paused: 0
 Stopped: 19
Images: 328
Server Version: 1.12.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 381
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: active
 NodeID: en2leu93wlp34nkla3yxmyfos
 Is Manager: true
 ClusterID: eujs1t26dcn9io4jfb61y37gh
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.31.1.100
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-38-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859 GiB
Name: hubnet-1
ID: CG3A:5KVP:ITEW:CF74:GDPB:GLXU:PK5D:5Z6P:PXWD:AKAS:GKWK:D6LV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: j0hnsmith
Registry: https://index.docker.io/v1/
Labels:
 provider=generic
Insecure Registries:
 127.0.0.0/8
```

### PR DESCRIPTION
**Pull request by thaJeztah**
fixes https://github.com/docker/docker/issues/29901
fixes https://github.com/docker/docker/issues/29203

when listing images with `docker images`, and filtering
the list by `image:tag`, the digest was not set on images.

This change always sets the digest on images (if present),
so that it will be shown in the results, even when filtering.

Similarly, when filtering by `digest`, the image's tag
was not shown, resulting in a different presentation,
depending on the filter that was set.

Before this patch:

    docker images --digests alpine:latest
    REPOSITORY          TAG                 DIGEST              IMAGE ID            CREATED             SIZE
    alpine              latest              <none>              88e169ea8f46        8 days ago          3.98 MB

    docker images --digests alpine@sha256:dfbd4a3a8ebca874ebd2474f044a0b33600d4523d03b0df76e5c5986cb02d7e8
    REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
    alpine              <none>              sha256:dfbd4a3a8ebca874ebd2474f044a0b33600d4523d03b0df76e5c5986cb02d7e8   88e169ea8f46        8 days ago          3.98 MB

After this patch:

    docker images --digests alpine:latest
    REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
    alpine              latest              sha256:dfbd4a3a8ebca874ebd2474f044a0b33600d4523d03b0df76e5c5986cb02d7e8   88e169ea8f46        8 days ago          3.98 MB

    docker images --digests alpine@sha256:dfbd4a3a8ebca874ebd2474f044a0b33600d4523d03b0df76e5c5986cb02d7e8
    REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
    alpine              latest              sha256:dfbd4a3a8ebca874ebd2474f044a0b33600d4523d03b0df76e5c5986cb02d7e8   88e169ea8f46        8 days ago          3.98 MB

/cc @vdemeester @dmcgowan @tonistiigi

### CODE REVIEW

File: daemon/images.go [hunk 1]
```diff
@@ -129,28 +128,34 @@ func (daemon *Daemon) Images(imageFilters filters.Args, all bool, withExtraAttrs
...
+			if _, ok := ref.(reference.Canonical); ok {
+				newImage.RepoDigests = append(newImage.RepoDigests, ref.String())
+			}
+			if _, ok := ref.(reference.NamedTagged); ok {
+				newImage.RepoTags = append(newImage.RepoTags, ref.String())
+			}
+			if !didMatch && imageFilters.Include("reference") {
 				var found bool
 				var matchErr error
 				for _, pattern := range imageFilters.Get("reference") {
```
Comment (thaJeztah) [inline daemon/images.go]: I was actually looking at this; Isn't out behavior that if the _same_ filter is provided twice, that all provided filters should match? i.e. `--filter reference=foo:latest --filter reference=bar:latest` should only match images that have _both_ `foo:latest` and `bar:latest`?

Compared this with labels;

```
$ docker volume create --label one --label two foo
foo

$ docker volume create --label one bar
bar

$ docker volume ls --filter label=one --filter label=two
DRIVER              VOLUME NAME
local               foo
```

**Comment by thaJeztah** [comment]
oh, I see what's happening; working on a fix

**Comment by thaJeztah** [comment]
Changed to "WIP". Haven't had time this weekend to complete my changes

**Comment by dmcgowan** [comment]
The updated logic looks correct to me. I wonder if we still need to do filter on the cli side to filter out non-requested tags though.

For example
```
$ docker images --digests registry:2.1
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
registry            2.1                 sha256:e641943a78a8f634c16ad69f5c9d779f470b147865c2121d89c52ea0da6fc1bd   52bb991b482e        14 months ago       220 MB
$ docker images --digests registry:2.2                                                                    
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
registry            2.2                 sha256:30adb707d1b4d2ad694c38da3ea1e7d303fdbdce2538ab0372afe6f1523ae3c8   ad379b517aa6        11 months ago       225 MB
registry            2.2                 sha256:a9128f456da044c4132c41973b065d2e2c46b04682c65f23dbb5d0c51071d595   ad379b517aa6        11 months ago       225 MB
registry            2.2.1               sha256:30adb707d1b4d2ad694c38da3ea1e7d303fdbdce2538ab0372afe6f1523ae3c8   ad379b517aa6        11 months ago       225 MB
registry            2.2.1               sha256:a9128f456da044c4132c41973b065d2e2c46b04682c65f23dbb5d0c51071d595   ad379b517aa6        11 months ago       225 MB
```

This might be related to the integration test failures. I think in that case returning 2 entries for the different digests is correct but don't see the point in showing the extra tags on the cli.

**Comment by thaJeztah** [comment]
Yes, we need it for situations where (e.g.) we match on digest (which results in the `tag` being omitted), and there were some other cases.

**Comment by LK4D4** [comment]
@thaJeztah still WIP?

**Comment by dnephin** [comment]
Needs a rebase and CI is failing

**Comment by vdemeester** [comment]
ping @thaJeztah @tonistiigi 👼 🙏

**Comment by dnephin** [comment]
@thaJeztah Thanks for working on this!

Looks like this PR is inactive so I'm going to close it for now.

**Pull request closed without merging**